src/govuk/components/all.test.js

```diff
@@ -43,6 +43,14 @@ it('_all.scss renders to CSS without errors', () => {
 })
 
 describe.each(allComponents)('%s', (component) => {
+  let percySnapshot
+
+  beforeAll(() => {
+    // Support fetch() detection, upload via WebSocket()
+    global.window = { fetch, WebSocket }
```

Comment:
@36degrees To answer the "Why do we need this?"

For context, Jest used to run with **jsdom** by default

**Performance improvements**
https://jestjs.io/blog/2022/04/25/jest-28
https://jestjs.io/blog/2021/05/25/jest-27

Jest now runs all tests via a Node.js environment. No browser globals

The problem? Percy code is required (by Jest) as a ["browserified" bundle](https://github.com/percy/cli/blob/90cd2f3a5d83f9835fcb975751cb828262fb95f0/packages/core/package.json#L22), so hits this code path:
https://github.com/percy/cli/blob/19f8be4903543544bdef197f2f7d5a6a648ed37d/packages/sdk-utils/src/logger.js#L74

So this _restores_ both browser globals and fixes:

1. **Fetch API** for Percy to see if it's CLI agent is running
2. **WebSocket** for Percy to upload screenshots securely

Reply:
I have been wondering about splitting the Percy snapshotting out as it's not really part of the 'tests' and doesn't rely on anything else from Jest.

I suspect we might also need to do this if we wanted to conditionally run the visual regression tests only if certain files have changed.

Until then, to stop us having to go down that rabbit hole too far, would another way to solve this be to split the snapshotting out into its own Jest test case?

Reply:
@36degrees Yeah, appreciate this is a quick fix. Did see the new projects list?

```console
npx jest --selectProjects "JavaScript component tests"
```

Likewise we can run everything except a certain project:

```console
npx jest --ignoreProjects "Snapshot tests"
```

Reply:
Does Percy inherently need both jsdom and puppeteer because we're running it in Jest? Or is this a symptom of the fact we have these tests inside a file with a lot of other ~unrelated tests and some of them need jsdom and some need puppeteer?

Reply:
It's because we use [@percy/puppeteer](https://www.npmjs.com/package/@percy/puppeteer) which crashes without `window.fetch` and `window.WebSocket`. For performance reasons, Jest CLI no longer defaults to a simulated browser (jsdom) environment.

Here's an extract of their code ([via browserify bundle](https://github.com/percy/cli/blob/master/packages/sdk-utils/package.json#L23)) showing `.call(window)`

```mjs
(function() {
  /* Percy internals */
}).call(window);

if (typeof define === "function" && define.amd) {
  define("@percy/sdk-utils", [], () => window.PercySDKUtils);
} else if (typeof module === "object" && module.exports) {
  module.exports = window.PercySDKUtils;
}
```

I'm happy with this fix until we can investigate calling [`@percy/client`](https://github.com/percy/cli/tree/master/packages/client#percyclient) directly

This PR is just to get `jest --watch` stable again (starting/stopping test site cleanly etc)

Reply:
To be honest, looks like a repeat of https://github.com/alphagov/govuk-frontend/issues/2643 again

>**[Jest v28 `package.json` `exports` support](https://jestjs.io/docs/28.x/upgrading-to-jest28#packagejson-exports)**
>…As a result, you might e.g. get browser code which assumes ESM, when Jest provides ['require', 'browser']. You can either report a bug to the library…

With Jest (correctly??) now loading the _default_ browser bundle for [@percy/puppeteer](https://www.npmjs.com/package/@percy/puppeteer) on this line:
https://github.com/percy/cli/blob/master/packages/sdk-utils/package.json#L27

```json
"exports": {
  ".": {
    "node": "./dist/index.js",
    "default": "./dist/bundle.js"
  }
}
```

We'll have to stick with this fix and consider a PR with Percy to add `import` and `require` exports